walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,170 @@
+   *
+   * @param events all events happening on a given day including event name, 
+   *                a TimeRange, and collection of attendees.
+   * @param request details of meeting being requested including requested meeting name, 
+   *                duration, and collection of attendees.
+   * @return an ArrayList containing all the TimeRanges the request meeting can be scheduled
+   *
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<String> attendees = new ArrayList<String>(request.getAttendees());
+    List<TimeRange> busyTimeRanges = findBusyTimes(events, attendees);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      List<String> optional = new ArrayList<String>(request.getOptionalAttendees());
+      attendees.addAll(optional);
+
+      List<TimeRange> optionalBusyTimeRanges = findBusyTimes(events, attendees);
+      List<TimeRange> optionalMeetingTimes = findMeetingTimes(optionalBusyTimeRanges, duration);
+
+      // Return optionalMeetingTimes.
+      if(!optionalMeetingTimes.isEmpty()) {
+        return optionalMeetingTimes;
+      } else if (request.getAttendees().isEmpty()){
+        return Arrays.asList();
+      }
+    }
+
+    return findMeetingTimes(busyTimeRanges, duration);
+  }
+
+  // Finds meeting times that don't fall during busy times of the day.
+  private List<TimeRange> findMeetingTimes(List<TimeRange> busyTimeRanges, long duration) {
+    if (busyTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(busyTimeRanges, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> busyTimeMerged = mergeBusyTimes(busyTimeRanges);
+    List<TimeRange> sortByEndTime = sortByEnd(busyTimeMerged);
+
+    return findFreeTimes(busyTimeMerged, sortByEndTime, duration);
+  }
+
+  // Merges all the overlapping busy times for all event attendees to create larger time interval, if possible.
+  private List<TimeRange> mergeBusyTimes(List<TimeRange> busyTimeRanges) {
+    List<TimeRange> merged = new ArrayList<TimeRange>();
+
+    int numRange = busyTimeRanges.size();
+    if (numRange == 1) { //only one busy time range, no need to merge
+      return busyTimeRanges;
+    }
+ 
+    int currRangeIndex = 0;
+    int nextRangeIndex = 1;
+    int start = busyTimeRanges.get(0).start();
+    int end = busyTimeRanges.get(0).end();
+
+    while (currRangeIndex < numRange && nextRangeIndex < numRange) {
+      TimeRange currRange = busyTimeRanges.get(currRangeIndex);
+      TimeRange nextRange = busyTimeRanges.get(nextRangeIndex);
+
+      if (currRange.contains(nextRange)) {
+        nextRangeIndex++;
```

Comment:
I think you want to add a "continue;" after incrementing the index. Otherwise you will add a time to the merged list, but in this case we don't want to.

Reply:
But what if the nextRange contains the range after the nextRange, then I would not be testing for that if I put continue... I just put in a test to print out all the options and it doesn't include the nextRange it if's contained in currRange so this seems to work but I could also be wrong.

Reply:
You are right. Instead you probably want

while (currRange.contains(nextRange) && nextRangeIndex + 1 < numRange) {
        nextRangeIndex++; 
)

Reply:
ok got it!

Reply:
I was wrong. I thought add was within the while loop, but it is outside the while loop so this should be good.